DEV9: Add HDD emulation

TheLastRar opened the pull request:
This is mostly the code from CLR_DEV9, with some tweaks and changes that I spotted during the porting process.

While CLR_DEV9 had split the speed emulation into a separate file, I chose to not do so for bringing in my changes to the speed code.

The emulator will auto create the file if it is missing

As a todo list

- [x] Add to Linux build scripts
- [x] Wire up the config for Linux
- [x] Config HDD UI for windows
- [x] Config HDD UI for Linux
- [x] HDD creation progress UI
- [x] Mac OS support (Requires #3999 to be merged first)

RedDevilus commented on pcsx2/DEV9/ATA/HddCreate.cpp:
```diff
@@ -0,0 +1,108 @@
+
+void HddCreate::SetDone()
+{
+	fprintf(stdout, "%i / %i MiB\n", neededSize, neededSize);
```
Won't this print the same value double? Or will it differ in multiples or 1000/1024?

TheLastRar replied:
Progress is normally printed in SetFileProgress(), which is limited to output once a second
Because of this, the last progress update might get dropped (even though the file is fully written), so I added an extra log that just listed it as being fully written.

It's kindof a nitpick on my part, as It would also log "Done" anyway

I can remove that one line if you prefer, I'm not fussed either way, it was kindof a place holder for the lack of a UI progress bar

RedDevilus replied:
I didnt understand why, thanks for explaining. It's fine as is for now.

TheLastRar commented on pcsx2/DEV9/ATA/ATA.h:
```diff
@@ -0,0 +1,269 @@
+	void HDD_Nop();
+	void HDD_Idle();
+
+	void DRQCmdPIODataToHost(u8* buf, int buffLen, int buffIndex, int size, bool sendIRQ);
```
No relation, I'm just inconsistent with shortening the word buffer

TheLastRar commented on pcsx2/DEV9/ATA/ATA.h:
```diff
@@ -0,0 +1,269 @@
+	void HDD_ReadPIO(bool isLBA48);
+	void HDD_ReadPIOS2();
+	void HDD_ReadPIOEndBlock();
+	//HDd_Write*
```
Woops

RedDevilus replied:
Yeah there is Hdd inconsistenties, but thats just some nitpicks

TheLastRar replied:
I think HddCreate is the main inconsistency (which I can easily fix if desired)

I also have variables such as hddImage, but that is lowercase because of variable naming, not sure what else

RedDevilus replied:
Well if it's distinct and somewhat consistent, i'm fine with it. I don't want to damage actual function over some formatting but it was more for readability in the first place.

TheLastRar commented:
I'll get to work on getting Linux to build sometime tomorrow

It seems like the Linux xml already saves/loads the Hdd settings, so that shouldn't need any extra work in that department

TheLastRar commented:
I believe I have all the UI bits done now

tadanokojin commented:
Tested Dark Chronicle NTSC-J for about 30 mins didn't run into any issues.

ichee commented:
Metal Saga NTSC-U working fine as well.

Also, if the HDD is enabled without specifying a path, then the emulator will crash after the PlayStation 2 screen.

So maybe having a default path pre-configured, such as the 'memcards' folder, could prevent this issue.

[emuLog.txt](https://github.com/PCSX2/pcsx2/files/5670425/emuLog.txt)

TheLastRar commented:
Added checks to disable HDD emulation if no path is specified

Adjusted the windows UI to require you to specify a path if HDD is enabled
The linux UI didn't get this check added, but it also lacks a similar check for selecting an ethernet device.

ichee commented:
Requiring a path to be specified before enabling is fine.
Configs should stay in the inis folder for consistency.

TheLastRar commented:
Rebased to resolve conflicts with the PAD merge, but I'm still waiting on https://github.com/PCSX2/pcsx2/pull/3999 for getting mac to build

TheLastRar commented:
Rebased again to resolve merge conflicts, also MAC OSX support is ready (I forgot to mention that here)

GovanifY commented on pcsx2/DEV9/ATA/ATA_Info.cpp:
```diff
@@ -0,0 +1,391 @@
+	*(u16*)&data[*index] = value;
+	*index += sizeof(value);
+}
+void WriteUInt32(u8* data, int* index, u32 value)
```
Maybe linefeed between functions?

GovanifY commented on pcsx2/DEV9/ATA/ATA_Info.cpp:
```diff
@@ -0,0 +1,391 @@
+	const u16 sectorSize = 512;
+	DEV9_LOG_VERB("HddSize : %i\n", config.HddSize);
+	const u64 nbSectors = ((u64)(sizeMb / sectorSize) * 1024 * 1024);
+	DEV9_LOG_VERB("nbSectors : %i\n", nbSectors);
```
Please use Console.Devlog for things like this

refractionpcsx2 replied:
Do you mean DevCon.WriteLn?

GovanifY replied:
That's what I meant indeed, brainfart

TheLastRar replied:
Is this just for those lines? or elsewhere?

GovanifY replied:
There are quite a few occurences in DEV9, I suggest you to do a `grep -nr DEV9_LOG` to find all of them or equivalent

TheLastRar replied:
Done, I've also replaced any emu_printf/printf/fprintf into Console.WriteLn/Error as I thought necceserry

GovanifY commented on pcsx2/DEV9/ATA/ATA_Info.cpp:
```diff
@@ -0,0 +1,391 @@
+	*(u32*)&data[*index] = value;
+	*index += sizeof(value);
+}
+void WriteUInt64(u8* data, int* index, u64 value)
```
Why not encapsulate them into a DEV9 namespace?

TheLastRar replied:
As in a `namespace DEV9` block? or something else? 

Didn't think anything in DEV9 was in a namespace

GovanifY replied:
Ah right, they're not because I didn't get around to redesign the plugin API yet. Can you move them in the ATA namespace, at least temporarily?

TheLastRar replied:
I'll stick it into the class, as there isn't a ATA namespace either

GovanifY commented on pcsx2/DEV9/ATA/ATA_Info.cpp:
```diff
@@ -0,0 +1,391 @@
+	WriteUInt16(identifyData, &index, curSectors); //word 56
+	//Current capacity in sectors
+	WriteUInt32(identifyData, &index, (u32)curOldsize); //word 57-58
+	//M
```
typo? happens multiple times

TheLastRar replied:
The `//M` comment? It's to to with ATA Flash memory (M-Flash in the standards iirc), which I'm not supporting here.
I would need to refresh myself on what the full name of the field is.

TheLastRar replied:
Nope, the M was something I did to keep track of something, but forget what is was I put it in for, will be amending that

GovanifY commented on pcsx2/DEV9/ATA/ATA_Transfer.cpp:
```diff
@@ -0,0 +1,250 @@
+	if (hddImage.fail())
+	{
+		DEV9_LOG_ERROR("Write error");
+		assert(false);
```
Use pxAssert when possible

GovanifY reviewed (changes requested):
Few things to change here and there, the actual logic sounds sound unless I missed something stupid. If testing works along with those changes I'm fine with merging it

TheLastRar commented:
I believe that is all of @GovanifY comments taken care of

refractionpcsx2 commented:
> I believe that is all of @GovanifY comments taken care of

hey dude, can you have a quick look on Discord, I've made some comments relating to some of this, might require a small amount of discussion and don't want to spam the github with it, cheers :)

TheLastRar commented:
Squashed down the commits, hopefully I didn't fuck anything up.

Is there more that you want squashed down?
